Bugfix for multi-range scan with MaxResults parameter

joezxy opened the pull request:
In existing code, when dist_sender processes one multi-range ScanRequest, it will dispatch the request with MaxResults parameter to each range one by one, and the return results will be combined and returned. The problem is if there are 2 ranges, MaxResults is set to 2, so each individual ScanRequest will return 2 rows, and finally 4 rows will be returned, it does not satisfy the MaxResults parameter "2". and if ScanRequest in range1 already got enough results, it need not to access range2.

Because dist_sender should focus on the dispatching mechanism, directly putting ScanRequest/ScanReponse into dist_sender seems not good, so I add two interfaces into api.go, please check whether it is ok. Thanks.

joezxy committed:
Bugfix for multi-range scan with MaxResults parameter
Remove unnecessary type assertion

tschottdorf commented on kv/dist_sender.go:
```diff
@@ -396,6 +403,15 @@ func (ds *DistSender) Send(call *client.Call) {
 			} else if isMulti {
 				// If this request spans ranges, collect the replies.
+				if maxResults > 0 {
+					if reply, ok := reply.(proto.Truncatable); ok {
```
Maybe we can manage to do all the multi-range work here, so you could Combine() here, then do a RowCount() (if applicable) and stop quering more ranges (or truncate & break if you're above maxRetries)
Ideally that should be done without allocating extra variables when the request does not span ranges.

I also think there are some bugs hidden here due to `gogoproto.Merge` (see https://github.com/cockroachdb/cockroach/issues/494) without properly resetting. Feel free to fix those when you spot them.

tschottdorf replied:
Added: Don't let the #494 stuff distract you from the main point of the PR - I just wanted to mention it. Looks like the proper fix for that is actually upstream in txn_coord_sender, so it should probably be a separate PR anyways.

joezxy replied:
Thanks for your comments, @tschottdorf , one small suggestions: If we put both maxResults and accumulateResults into the "isMulti" branch, then each time when accessing one range inside a multi-range request, accumulatedResults needs to be calculated again according to the received responses, some performance burden, right? maybe it is better to keep accumulatedResults outside.

tschottdorf commented on proto/api.go:
```diff
@@ -632,3 +632,33 @@ func (br *BatchResponse) Add(reply Response) {
+	}
+}
+
+// RowCount returns the length of rows in ScanResponse.
```
s/length/number of/

tschottdorf commented on proto/api.go:
```diff
@@ -632,3 +632,33 @@ func (br *BatchResponse) Add(reply Response) {
+}
+
+// Truncatable is implemented by response types whose corresponding
+// requests has bounded result rows, such as Scan.
```
s/has/have/

tschottdorf commented on proto/api.go:
```diff
@@ -632,3 +632,33 @@ func (br *BatchResponse) Add(reply Response) {
 	br.Responses = append(br.Responses, union)
 }
+
+// Bounded is implemented by request types which has bounded result rows,
```
s/has bounded/have a bounded number of/

joezxy commented:
Addressed the comments, @tschottdorf , could you take another look?

spencerkimball commented on kv/dist_sender.go:
```diff
@@ -397,6 +397,50 @@ func (ds *DistSender) Send(call *client.Call) {
 			} else if isMulti {
 				// If this request spans ranges, collect the replies.
 				responses = append(responses, reply)
+
```
This method has gotten rather large. I'd love to find a way to refactor it into more manageable pieces. Could you add a general TODO here?

spencerkimball commented on kv/dist_sender.go:
```diff
@@ -397,6 +397,50 @@ func (ds *DistSender) Send(call *client.Call) {
+				// Then combine and truncate the responses.
+				if descNext == nil {
+					// Combine multiple responses into the first one.
+					if len(responses) > 1 {
```
Perhaps make this slightly more terse:

    if descNext == nil {
      for _, r := range responses[1:] {
        responses[0].(proto.Combinable).Combine(r)
      }
      ....
    }

spencerkimball commented on kv/dist_sender.go:
```diff
@@ -397,6 +397,50 @@ func (ds *DistSender) Send(call *client.Call) {
+					}
+
+					// Truncate the response if MaxResults is specified.
+					if maxResults > 0 {
```
Another option which might be slightly more efficient is to keep track of the running total in the outer loop (i.e. move accumulatedResults var to a higher-level scope) and provide a SetBound() method on the proto.Bounded interface for each call through to a range.

spencerkimball commented:
LGTM

joezxy commented:
Addressed the comments, @spencerkimball , for the 3rd comment, because moving accumulatedResult out of the loop already avoids to count the row number for each range repeatedly, it might not be necessary to expose SetBound() through interface any more. I think maybe we can add this method if in future there are some scenarios which need to change the Bound parameter dynamically.

bdarnell commented on kv/dist_sender.go:
```diff
@@ -397,6 +402,45 @@ func (ds *DistSender) Send(call *client.Call) {
+					maxResults = args.GetBound()
+					if reply, ok := reply.(proto.Truncatable); ok {
+						// Calculate the number of the existing rows.
+						accumulatedResults += reply.(proto.Truncatable).RowCount()
```
You can simply use `reply.RowCount()` here; `reply` is already a `proto.Truncatable` in this scope.

bdarnell commented:
LGTM

spencerkimball commented:
My idea with providing SetBound() is it would allow us to specify the correct remaining max number of results for calls to ranges after the first, possibly resulting in less network traffic and removing the requirement that we truncate any results.

LGTM

spencerkimball commented:
Oops didn't mean to close that.

joezxy commented:
Got your point, @spencerkimball ,opened another issue for this enhancement.
